### ISSUE TITLE
Feature request: API to allow tied computation of function value and gradient

### ISSUE THREAD

**Issue by bsxfan** [JuliaNLSolvers/Optim.jl#701]
Many (maybe most) non-trivial objective functions are most naturally coded to share some computational resources when computing the function value and the gradient. But, unless I'm missing something, the API for `Optim` does not support this, because it requires completely separate functions for value and gradient.

A very natural way to make a gradient available for some function, `obj:R^n->R` is as follows:
```
value,g = obj(x)
gradient = g(1.0)  # This is the gradient at x.  (The argument of g() allows backpropagation
                             #  of derivates when combining obj with some outer function. )
```
The first invocation of `obj(x)` computes the value at `x` and also wraps in the lambda function `g` whatever intermediate results it needs to more efficiently compute the gradient at `x`.

Is there any way that the existing API for `Optim` would already allow such sharing of resources between value and gradient, or would it need an API re-design?

In some experiments with the LBFGS solver (with default line search) I noticed it calls gradient and value like this:
```
gradient(x1)
value(x1)
gradient(x2)
value(x2)
gradient(x3)
value(x3)
... etc ...
```
If one makes the fragile assumption that this is always the case, one can kludge together a more efficient solution where `gradient(x)` computes both value and gradient, caches the value somewhere and returns the gradient. Then  value(x) ignores its input and just returns the cached value.

Is there a nicer and more robust way to do this?

**Comment by antoine-levitt** [comment]
That is sort-of-documented in https://julianlsolvers.github.io/Optim.jl/stable/#user/tipsandtricks/, but this is a very bad API. A better option might be using the `OnceDifferentiable(f, g!, fg! x)` constructor for your objective function.

**Comment by pkofod** [comment]
Oh, I thought I had tagged a new release. This is already possible, but not really documented. I have yet to add a convenience API for it, but you can do it as stated in the docs for the master branch https://julianlsolvers.github.io/Optim.jl/latest/#user/tipsandtricks/ (@antoine-levitt is this still bad to you?)

**Comment by pkofod** [comment]
To be quite specific, all the first order methods(!) in Optim (GradientDescent, ConjugateGradient, (L)BFGS, the accelerated methods and NGMRES) already have an internal API where they call a function called `value_gradient!` to update the results, though notice that it's unfortunately hidden in a function called `update_g!` :confused: 
https://github.com/JuliaNLSolvers/Optim.jl/blob/d7b052ea2f5bde0e72ebcbaf6f91effd4632da5a/src/multivariate/optimize/optimize.jl#L4

This calls into NLSolversBase's OnceDifferentiable type. What this does is basically it calculates only objective if the gradient has already been evaluated here, it calculates only the gradient if the objective has already been evaluated here and it calculates the value and gradient simultaneously if neither has been evaluated at the current iterate. If you pass in a OnceDifferentiable type constructed like @antoine-levitt showed above, then you get the fully performant call in the line I posted above. If you just passed in `f` and `g!`, then the "simultaneous" evaluation is basically a sequence of calls: `f(x)` followed by `g!(G, x)`.

However, if you use the way of doing it that I showed you in my previous comment, then you actually get the fully efficient evaluation given the circumstance when evaluating the two together.

Now, what we cannot do here, with our current api, is to do what you want with sharing states through a lambda. That would require a significant rewrite, and/or a separate API for that. Maybe we could allow the user to provide a `setup_vars`-ish function that would then update some variables that `fg!` closes over.

So to sum up, we cannot currently provide an API where `f` and `g!` share calculations that should be caches between calls, but as long as they're called **together**, we actually have an API for that. Most calls, besides those in the line search iterations, actually end up trying to evaluate the two together, so if you follow the directions in the documentation for the master branch, then it should be *quite*  efficient, but maybe not 100% optimal. For that, you'd have to do something like the "ugly" trick Antoine mentioned.

**Comment by bsxfan** [comment]
Thank you @antoine-levitt and @pkofod for your help. I have managed to get the optimizer to work according to my requirements by using your suggestions. What I did can be summarized as:
```
using Optim
fg!(flag,G,x) = compute_both!(x,G)
fg!(flag::Nothing,G,x) = compute_gradient!(x,G)
fg!(flag,G::Nothing,x) = compute_value(x)
obj = Optim.only_fg!(fg!)
result = optimize( obj, x0, LBFGS() )
```
In this case, LBFGS (with default line-search) only ever invokes the `compute_both!` variant.

**Comment by pkofod** [comment]
Glad to hear it worked!
